File: 1/1000s/13s/1353/MaximumNumberOfEventsThatCanBeAttended.cpp
/*
https://leetcode.com/problems/maximum-number-of-events-that-can-be-attended/description/?envType=daily-question&envId=2025-07-07
*/

class Solution {
public:
    int maxEvents(vector<vector<int>>& events) {
        sort(events.begin(), events.end());
        priority_queue<int, vector<int>, greater<>>pq;
        int sz=events.size(), idx=0, maxDay=0, res=0;
        for(auto &event: events){
            maxDay=max(maxDay, event[1]);
        }
        for(int curDay=1; curDay<=maxDay; ++curDay){
            while(idx<sz&&events[idx][0]==curDay){
                pq.push(events[idx++][1]);
            }
            while(!pq.empty()&&pq.top()<curDay){
                pq.pop();
            }
            if(!pq.empty()){
                pq.pop();
                res++;
            }
        }
        return res;
    }
};

File: 1/1000s/13s/1353/MaximumNumberOfEventsThatCanBeAttended.java
/*
https://leetcode.com/problems/maximum-number-of-events-that-can-be-attended/description/?envType=daily-question&envId=2025-07-07
*/

class Solution {
    public int maxEvents(int[][] events) {
        Arrays.sort(events, (A, B)->Integer.compare(A[0], B[0]));
        PriorityQueue<Integer>minHeap=new PriorityQueue<>((A, B)->(A-B));
        int lengthOfEvents=events.length, beginIndex=0, maxDay=0, finalMaxEvents=0;
        for(int[] event: events){
            maxDay=Math.max(maxDay, event[1]);
        }
        for(int curDay=1; curDay<=maxDay; ++curDay){
            while(beginIndex<lengthOfEvents&&events[beginIndex][0]==curDay){
                minHeap.add(events[beginIndex++][1]);
            }
            while(!minHeap.isEmpty()&&minHeap.peek()<curDay){
                minHeap.poll();
            }
            if(!minHeap.isEmpty()){
                minHeap.poll();
                finalMaxEvents++;
            }
        }
        return finalMaxEvents;
    }
}